# Importing data and libraries

In [1]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

11501568/11490434 [==============================] - 0s 0us/step


In [3]:
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

# Task 1. In this task, you'll implement several ANN models with different activation functions. For each, use the cross-entropy loss function as the loss function

## 1. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [4]:
model = Sequential()
# our first dense layer is 128
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# our second dense layer is 64
model.add(Dense(64, activation="tanh"))
# last layer, output layer, is 10
model.add(Dense(10, activation="softmax"))

# using categorical crossentropy as the loss
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0240 - accuracy: 0.7492
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5155 - accuracy: 0.8720
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4172 - accuracy: 0.8900
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3703 - accuracy: 0.8997
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3415 - accuracy: 0.9053
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3210 - accuracy: 0.9102
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3053 - accuracy: 0.9136
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2923 - accuracy: 0.9176
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2814 - accuracy: 0.9207
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2717 - accuracy: 0.9230

In [5]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.20555536448955536
Test accuracy: 0.9416000247001648


## 2. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [6]:
model = Sequential()
# our first dense layer is 128
model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
# our second dense layer is 64
model.add(Dense(64, activation="sigmoid"))
# last layer, output layer, is 10
model.add(Dense(10, activation="softmax"))

# using categorical crossentropy as the loss
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 2.2806 - accuracy: 0.1877
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 2.2196 - accuracy: 0.3766
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 2.1482 - accuracy: 0.5003
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 2.0472 - accuracy: 0.5575
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 1.9052 - accuracy: 0.5942
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 1.7259 - accuracy: 0.6209
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 1.5351 - accuracy: 0.6539
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 1.3602 - accuracy: 0.6858
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 1.2118 - accuracy: 0.7180
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0896 - accuracy: 0.7434

In [7]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.5503221750259399
Test accuracy: 0.858299970626831


## 3. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [8]:
model = Sequential()
# our first dense layer is 128
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer is 64
model.add(Dense(64, activation="relu"))
# last layer, output layer, is 10
model.add(Dense(10, activation="softmax"))

# using categorical crossentropy as the loss
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 1.2335 - accuracy: 0.6851
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5055 - accuracy: 0.8682
Epoch 3/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3874 - accuracy: 0.8935
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3396 - accuracy: 0.9042
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3110 - accuracy: 0.9116
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2904 - accuracy: 0.9174
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2737 - accuracy: 0.9218
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2600 - accuracy: 0.9261
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2481 - accuracy: 0.9303
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2373 - accuracy: 0.9328

In [9]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.1686592996120453
Test accuracy: 0.9503999948501587


## 4. Compare the results of each model. Which activation function performed best?

Loss = Cross Entropy
- activation = 'tanh'
 - train accuracy 0.9414
 - train score 0.2066
 - test accuracy 0.9416
 - test score 0.2056
- activation = 'sigmoid'
 - train accuracy 0.8543
 - train score 0.5739
 - test accuracy 0.8583
 - test score 0.5503
- activation = 'relu'
 - train accuracy 0.9528
 - train score 0.1669
 - test accuracy 0.9504
 - test score 0.1687

The relu activation method worked the best with the train and est scores being very similar so we know there is probably not any overfitting as well as having the highest accuracy and lowest loss score amongst the different activation methods tested

# Task 2. In this task, you'll implement the ANN models specified below. For each, use the hinge loss function as the loss function.

## 1. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [10]:
model = Sequential()
# our first dense layer is 128
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# our second dense layer is 64
model.add(Dense(64, activation="tanh"))
# last layer, output layer, is 10
model.add(Dense(10, activation="softmax"))

# using categorical crossentropy as the loss
model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 0.9884 - accuracy: 0.4326
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.7781 - accuracy: 0.6790
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5852 - accuracy: 0.7930
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4632 - accuracy: 0.8480
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3885 - accuracy: 0.8668
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3423 - accuracy: 0.8776
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3114 - accuracy: 0.8836
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2891 - accuracy: 0.8897
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2722 - accuracy: 0.8944
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2588 - accuracy: 0.8981

In [11]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.1859210580587387
Test accuracy: 0.9225999712944031


## 2. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [12]:
model = Sequential()
# our first dense layer is 128
model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
# our second dense layer is 64
model.add(Dense(64, activation="sigmoid"))
# last layer, output layer, is 10
model.add(Dense(10, activation="softmax"))

# using categorical crossentropy as the loss
model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0232 - accuracy: 0.0888
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0096 - accuracy: 0.1286
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0034 - accuracy: 0.1825
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0022 - accuracy: 0.2456
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0017 - accuracy: 0.3007
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0012 - accuracy: 0.3508
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0008 - accuracy: 0.3958
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0004 - accuracy: 0.4368
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0001 - accuracy: 0.4702
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 0.9997 - accuracy: 0.5019

In [13]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.9968549609184265
Test accuracy: 0.6859999895095825


## 3. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [14]:
model = Sequential()
# our first dense layer is 128
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer is 64
model.add(Dense(64, activation="relu"))
# last layer, output layer, is 10
model.add(Dense(10, activation="softmax"))

# using categorical crossentropy as the loss
model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0108 - accuracy: 0.3401
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.9743 - accuracy: 0.5845
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.8575 - accuracy: 0.6564
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.6502 - accuracy: 0.7725
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4666 - accuracy: 0.8494
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3591 - accuracy: 0.8730
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3047 - accuracy: 0.8846
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2725 - accuracy: 0.8925
Epoch 9/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2507 - accuracy: 0.8979
Epoch 10/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2346 - accuracy: 0.9032

In [15]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.15910187363624573
Test accuracy: 0.9319000244140625


## 4. Compare the results of each model with the result of the same model from the previous task. Which loss function performed best?

Loss = Cross Entropy, Hinge
- activation = 'tanh'
 - train accuracy 0.9414, 0.9185
 - train score 0.2066, 0.1936
 - test accuracy 0.9416, 0.9226
 - test score 0.2056, 0.1856
- activation = 'sigmoid'
 - train accuracy 0.8543, 0.6731
 - train score 0.5739, 0.9971
 - test accuracy 0.8583, 0.6856
 - test score 0.5503, 0.9976
- activation = 'relu'
 - train accuracy 0.9528, 0.9276
 - train score 0.1669, 0.1665
 - test accuracy 0.9504, 0.9319
 - test score 0.1687, 0.1591

The relu activation method worked the best in this series of models, but all of the different activation methods worked better using cross entrpy as the loss function vs hinge.